### ISSUE TITLE
Performance degradation for nightly-20240109

### ISSUE THREAD

**Issue by cyliu0** [risingwavelabs/risingwave#14468]
### Describe the bug

```
+-----------------------------------------------------------------+--------------+------------+---------------------+-----------------------------+
| BENCHMARK NAME                                                  | EXECUTION ID | STATUS     | FLUCTUATION OF BEST | FLUCTUATION OF LAST 10 DAYS |
+-----------------------------------------------------------------+--------------+------------+---------------------+-----------------------------+
| nexmark-q4-blackhole-medium-1cn-affinity                        |        18144 | Negative   | -26.91%             | -12.18%                     |
| nexmark-q20-blackhole-medium-1cn-affinity                       |        18156 | Negative   | -19.34%             | -11.74%                     |
| nexmark-q9-blackhole-medium-1cn-affinity                        |        18161 | Negative   | -30.70%             | -15.74%                     |
| sysbench-oltp-point-select-medium-3cn                           |        18165 | Negative   | -39.56%             | -21.76%                     |
| sysbench-select-random-points-medium-3cn                        |        18166 | Negative   | -32.88%             | -14.88%                     |
| sysbench-select-random-ranges-medium-3cn                        |        18168 | Negative   | -38.10%             | -18.27%                     |
| sysbench-oltp-insert-medium-3cn                                 |        18169 | Negative   | -33.20%             | -22.09%                     |
| sysbench-select-random-limits-medium-3cn                        |        18171 | Negative   | -58.32%             | -27.49%                     |
```

http://metabase.risingwave-cloud.xyz/dashboard/278-sysbench-rw-qps
http://metabase.risingwave-cloud.xyz/dashboard/277-nexmark-blackhole-affinity-1cn-rw-avg-source-throughput?namespace=daily

### The version of RisingWave

nightly-20240109

### Additional context

https://github.com/risingwavelabs/rw-commits-history?tab=readme-ov-file#nightly-20240109

**Comment by BugenZhao** [comment]
Still possibly caused by #14220 as there was an OOM issue on 8, Jan.

If this is the case, it will not be surprising that the sysbench performance degrades more significantly than streaming queries. This is because batch queries with high QPS result lead to a larger number of traces.

However, it's still interesting to find only 3 nexmark queries are affected. 🫨 I conducted a benchmark for some selected queries on that PR but found the performance impact is minimal. It could be coincidental.

Anyhow, I'll consider adding a runtime switch for this feature.

**Comment by xxchan** [comment]
> it's still interesting to find only 3 nexmark queries are affected.

This is also the most confusing fact to me.

Let's try to run the benchmarks with https://github.com/risingwavelabs/risingwave/pull/14220 reverted to confirm it first?

**Comment by BugenZhao** [comment]
> Let's try to run the benchmarks with #14220 reverted to confirm it first?

Does the benchmark pipeline accept env vars? Setting `RW_DISABLE_EMBEDDED_TRACING` would be good.

**Comment by xxchan** [comment]
IIRC not

**Comment by cyliu0** [comment]
Currently, the only way to do it quickly is adding the env var to kube-bench https://github.com/risingwavelabs/kube-bench/blob/main/manifests/risingwave/risingwave.template.yaml#L151 in a branch. And then we can run the pipeline by specifying the `KUBEBENCH_BRANCH` when we trigger the job

**Comment by xxchan** [comment]
BTW, I noticed that actually more than those queries have a regression. 🤔 (But still, not all)

<img width="1575" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/37948597/534f0b06-3db6-498b-92a2-0ff583ab35c0">

<img width="543" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/37948597/98559bde-2026-4c6b-a85a-338558dddf8a">

**Comment by st1page** [comment]
https://www.notion.so/risingwave-labs/Performance-degradation-for-nightly-20240109-1d5e15ebba3247a089ddabe78d5d668f

**Comment by st1page** [comment]
There is also a performance degradation between nightly-20240101 and 02 in q9, will take a look later
<img width="969" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/40830455/3ee92b7a-bbc6-4317-8a53-731cdca1741f">

**Comment by st1page** [comment]
> There is also a performance degradation between nightly-20240101 and 02 in q9, will take a look later <img alt="image" width="969" src="https://private-user-images.githubusercontent.com/40830455/295494197-3ee92b7a-bbc6-4317-8a53-731cdca1741f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4NzY2MDYsIm5iZiI6MTcwNDg3NjMwNiwicGF0aCI6Ii80MDgzMDQ1NS8yOTU0OTQxOTctM2VlOTJiN2EtYmJjNi00MzE3LThhNTMtNzMxY2RjYTE3NDFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTEwVDA4NDUwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAwY2Q2ZDBiYjQ2NTYxNTllM2M2ZTNiNWNjNzMwMDAzNzI4YjU5NzkwOWI0ZmFhNWViOGQ5YjcyYmQ2NDE5MzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.58TXi0YzMZQp0Rvn1nsGOXvuTluCcswZYCgF5fMLpjc">

https://github.com/risingwavelabs/risingwave/issues/14484

**Comment by st1page** [comment]
- `d21d42f9a8ab63a78f67241480c2009ec6f1f61b` [feat(sink): Support without backfill sink with 'create sink from mv'. (#13947)](https://github.com/risingwavelabs/risingwave/pull/13947)
- `37a666a709ef15f0e3747cda7ef9824daeec5e65` [fix(mysql-cdc): only read table schema of captured tables instead of all tables for shared source (#14406)](https://github.com/risingwavelabs/risingwave/pull/14406)
- `c6f572db1bbbc2c7ac975d2b0b87ca688d09dbef` [chore(meta): change default parallelism for already existed streaming job to Custom (#14443)](https://github.com/risingwavelabs/risingwave/pull/14443)
- `fd768daa31f1f7f79965f62eed1e47574f0b85e3` [fix(connector): avoid long-lived tracing span (#14434)](https://github.com/risingwavelabs/risingwave/pull/14434)
- `d20632da0ff56ce57113dafe67b73a8a1415cb4a` [feat: poll barrier complete rpc in barrier manager worker loop (#14419)](https://github.com/risingwavelabs/risingwave/pull/14419)
- `a30583f8c1edd3357b9985d172a551cad1c9cfd0` [chore: improve error message for function argument mismatch (#14435)](https://github.com/risingwavelabs/risingwave/pull/14435)
- `0799621be4161979f7d980921b471a45bf70d76b` [chore: bump `thiserror-ext` to 0.0.11 (#14439)](https://github.com/risingwavelabs/risingwave/pull/14439)
- `ef8efc79c0129e5ed3481de63f3781fc83c7c5f8` [fix: fix dyn filter's always relax workaround (#14433)](https://github.com/risingwavelabs/risingwave/pull/14433)
- `e6a4ffcb1edc2ddb7dfcec394f4d65d0e7444f5e` [fix(compactor): check compaction result for fast compact (#14274)](https://github.com/risingwavelabs/risingwave/pull/14274)
- `580a60ba72a32efdbcb95f029fcf7bccdf7b8d4c` [test: update `similar-to-escape` to test explicitly disable escape (#14393)](https://github.com/risingwavelabs/risingwave/pull/14393)
- `0c5abb429a6a136f95d4f2162a8c5eaff46a3db8` [chore(deps): Bump serial_test from 2.0.0 to 3.0.0 (#14420)](https://github.com/risingwavelabs/risingwave/pull/14420)
- `867f7751a01c4a9dcbfe01f18841e73308e467e1` [chore(deps): Bump zeroize from 1.6.0 to 1.7.0 (#14421)](https://github.com/risingwavelabs/risingwave/pull/14421)
- `babfdc9a4650e1b5d78a3676257b8dcc1f182e73` [chore: remove some periodical logs (#14427)](https://github.com/risingwavelabs/risingwave/pull/14427)
- `31a85682430e9bafec9b2deadee3412b107d2a22` [fix: missing update of upstream fragment and actor ids for table replacement (#14381)](https://github.com/risingwavelabs/risingwave/pull/14381)

**Comment by Little-Wallace** [comment]
e6a4ffcb1edc2ddb7dfcec394f4d65d0e7444f5e close fast compaction and this will cost more CPU resource in compactor node.

**Comment by BugenZhao** [comment]
> > Let's try to run the benchmarks with #14220 reverted to confirm it first?
> 
> Does the benchmark pipeline accept env vars? Setting `RW_DISABLE_EMBEDDED_TRACING` would be good.

Benchmark is done. 😕 It could possibly be related to the traces, but the impact does not seem significant.

`nightly-20240107` vs `nightly-20240109` vs `nightly-20240109 (disable embedded tracing)`

![image](https://github.com/risingwavelabs/risingwave/assets/25862682/2a74127f-ef9c-4fe4-b3ca-cb855a1bf3d5)
![image](https://github.com/risingwavelabs/risingwave/assets/25862682/bfbc92e4-ebe1-45e8-9227-40da0e266733)
![image](https://github.com/risingwavelabs/risingwave/assets/25862682/a937bc90-b293-429a-995e-126c7444b498)

**Comment by cyliu0** [comment]
nightly-20240110 benchmark result
```
+-----------------------------------------------------------------+--------------+------------+---------------------+-----------------------------+
| BENCHMARK NAME                                                  | EXECUTION ID | STATUS     | FLUCTUATION OF BEST | FLUCTUATION OF LAST 10 DAYS |
+-----------------------------------------------------------------+--------------+------------+---------------------+-----------------------------+
| nexmark-q7-blackhole-watermark-medium-1cn-affinity              |        18192 | Negative   | -34.83%             | -10.17%                     |
| sysbench-oltp-point-select-medium-3cn                           |        18210 | Negative   | -40.48%             | -20.21%                     |
| sysbench-select-random-points-medium-3cn                        |        18211 | Negative   | -34.91%             | -15.67%                     |
| sysbench-select-random-ranges-medium-3cn                        |        18213 | Negative   | -39.44%             | -17.92%                     |
| sysbench-oltp-insert-medium-3cn                                 |        18215 | Negative   | -34.40%             | -20.66%                     |
| sysbench-select-random-limits-medium-3cn                        |        18216 | Negative   | -50.70%             | -10.61%                     |
```

**Comment by st1page** [comment]
> [e6a4ffc](https://github.com/risingwavelabs/risingwave/commit/e6a4ffcb1edc2ddb7dfcec394f4d65d0e7444f5e) close fast compaction and this will cost more CPU resource in compactor node.

So should we add it back?

**Comment by lmatz** [comment]
just to be safe, maybe let `fast compaction` be tested for a longer period of time before enabling it by default?

**Comment by xxchan** [comment]
> Benchmark is done. 😕 It could possibly be related to the traces, but the impact does not seem significant.
> 
> `nightly-20240107` vs `nightly-20240109` vs `nightly-20240109 (disable embedded tracing)`

Q20's perf even decreased in your test

**Comment by xxchan** [comment]
In yesterday's result, some querys' perf increased a lot (e.g., q9,19,20), which is also strange.

**Comment by st1page** [comment]
> In yesterday's result, some querys' perf increased a lot (e.g., q9,19,20), which is also strange.

There are still some compactor's CPU bottleneck in nightly-0110
<img width="851" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/40830455/83200083-7f96-4859-b2c7-82d6c2215484">

**Comment by st1page** [comment]
discussion with @Li0k 
Although we think we have close the fast compaction, we still find the metrics of it in 0109 and 0110. @Li0k will do some invesitigation
 
<img width="837" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/40830455/93c93f50-c6c3-4f03-9fa5-474227225b8e">

**Comment by lmatz** [comment]
I think it may be because https://github.com/risingwavelabs/kube-bench/pull/349 enables fast compaction all of the testing pipelines instead of just the perf test one.. let me correct it

**Comment by st1page** [comment]
> I think it may be because [risingwavelabs/kube-bench#349](https://github.com/risingwavelabs/kube-bench/pull/349) enables fast compaction all of the testing pipelines instead of just the perf test one.. let me correct it
> 
> edit: or do you want to keep the status quo?

So the degradation does not because we have closed the fast compaction. we still need to find which PR caused it....

**Comment by xxchan** [comment]
Aren't we already sure it's caused by tracing (and earlier regression)? :eyes: Which query's regression are we talking about now?

**Comment by st1page** [comment]
> Aren't we already sure it's caused by tracing (and earlier regression)? 👀 Which query's regression are we talking about now?

Today‘s bench result looks well.. let's watch and wait for tomorrow's result.

**Comment by st1page** [comment]
> > > Let's try to run the benchmarks with #14220 reverted to confirm it first?
> > 
> > 
> > Does the benchmark pipeline accept env vars? Setting `RW_DISABLE_EMBEDDED_TRACING` would be good.
> 
> Benchmark is done. 😕 It could possibly be related to the traces, but the impact does not seem significant.
> 
> `nightly-20240107` vs `nightly-20240109` vs `nightly-20240109 (disable embedded tracing)`

Can you bench the sysbench test without traces? I am not sure if the degradation of streaming nexmark and serving sysbench are caused by the same reason.

**Comment by BugenZhao** [comment]
> Can you bench the sysbench test without traces? I am not sure if the degradation of streaming nexmark and serving sysbench are caused by the same reason.

Since #14528 got merged, distributed tracing is now off by default. Let's say the result of sysbench tonight.

**Comment by BugenZhao** [comment]
I believe the degradation of sysbench has been resolved by #14528.
